### PR TITLE
Collations proof of concept

### PR DESCRIPTION
**Pull request by dbatomic**
# Rough POC for collations in Spark

## High level changes
- Collation suite that test currently supported features (start with this file).
- Global `CollatorFactory`. This global table represents stores all available collations in the system. It can provide collation aware comparators and hashes to the `UTF8String` function. 
- `UTF8String` is extended with single integer that specifies collation. We could be even more aggressive and pack this integer into a short, or even a byte. Id represents cached comparator id that can be fetched from `CollatorFactory`.
- `UTF8String` respects this id for equality checks and compares.
- `StringType` class is extended with collationId field. We keep using existing `case object StringType` to mark default, utf8 binary, collation. By doing this, we should keep binary compatibility with previous versions. 
- Extending `PhysicalStringType` with `collationId` field.
- Support for aggregates, given that they currently rely on pure byte for byte comparison for group building.
- Support for merge join (hash based joins are TODO).
- ICU is used as collation library. ICU4j only exposes UTF16 APIs which means that we need to do UTF8 -> UTF16 conversion on every comparison which is very suboptimal. Alternative is to ICU4c through JNI which does support UTF8 APIs.
- Unit benchmark that covers UTF8 operations with and without collators. At this point measured diff between collated and uncollated comparison is ~12x.

## Supported features at this point:
- `collate` expression -> input string is casted to `StringType(collation)`.
- Collation rules are ICU collator based. Caller provides locale and strength (primary, secondary, tertiary, identical). E.g. `collate(input, 'sr-primary')` will collate input with Serbian locale that ignores both casing and accents. Secondary will ignore casing but respect accents and tertiary will respect both.
- `collation` expression -> returns collation name of given input.
- Support for basic operators (filters, aggregate, joins, views, inline tables etc.).
- Support for `CREATE TABLE(a STRING COLLATE x)` syntax.
- Basic parquet support by disabling filter pushdown.

Proper testing (and creating real test strategy is TBD).

TBD is (proper) parquet and delta support, different collation levels (column level, table level, database level) and much more extensive testing of other features.

Suggestion for reviewers of this POC is to start with `CollationSuite` and newly tests in `UTF8StringSuite` to get the gist of the changes in this PR.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala [hunk 1]
```diff
@@ -207,14 +208,20 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper with Join
...
+
   /**
    * To be able to prune partitions on a join key, the filtering side needs to
    * meet the following requirements:
    *   (1) it can not be a stream
    *   (2) it needs to contain a selective predicate used for filtering
    */
   private def hasPartitionPruningFilter(plan: LogicalPlan): Boolean = {
-    !plan.isStreaming && hasSelectivePredicate(plan)
+    !plan.isStreaming && hasSelectivePredicate(plan) && hasNoCollations(plan)
```
Comment (milastdbx) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala:224]: Was your intention to disable pruning altogether ?

This pruning is just for dynamic pruning.

There is also pruning happening during file listing in both static analysis and scan phy nodes

Comment (dbatomic) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala:224]: Yep, idea is to disable pruning altogether. At this point pruning part is still not stable and I also don't have any test coverage for pruning. You are right, need to take care of other "prunings" as well.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -1390,8 +1497,13 @@ public UTF8String copy() {
 
   @Override
   public int compareTo(@Nonnull final UTF8String other) {
-    return ByteArray.compareBinary(
-        base, offset, numBytes, other.base, other.offset, other.numBytes);
+    if (comparatorId == 0) {
```
Comment (stefankandic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1500]: what if the `other` has nonzero comparatorId, should we do something like `-1 * other.compareTo(this)`?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 3]
```diff
@@ -1401,10 +1513,17 @@ public int compare(final UTF8String other) {
   @Override
   public boolean equals(final Object other) {
     if (other instanceof UTF8String o) {
-      if (numBytes != o.numBytes) {
-        return false;
+      if (comparatorId == 0)
```
Comment (stefankandic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1516]: same as for the above comment

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ToStringBase.scala [hunk 4]
```diff
@@ -256,11 +258,11 @@ trait ToStringBase { self: UnaryExpression with TimeZoneAwareExpression =>
       // as strings. Otherwise, the casting is using `BigDecimal.toString` which may use scientific
       // notation if an exponent is needed.
       case _: DecimalType if useDecimalPlainString =>
-        (c, evPrim) => code"$evPrim = UTF8String.fromString($c.toPlainString());"
-      case StringType =>
+        (c, evPrim) => code"$evPrim = UTF8String.fromString($c.toPlainString(), $collationId);"
+      case _: StringType =>
         (c, evPrim) => code"$evPrim = $c;"
```
Comment (stefankandic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ToStringBase.scala:263]: this needs collationId as well, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ToStringBase.scala [hunk 5]
```diff
@@ -231,11 +232,12 @@ trait ToStringBase { self: UnaryExpression with TimeZoneAwareExpression =>
              |$evPrim = $buffer.build();
            """.stripMargin
         }
-      case pudt: PythonUserDefinedType => castToStringCode(pudt.sqlType, ctx)
+      case pudt: PythonUserDefinedType =>
+        castToStringCode(pudt.sqlType, ctx, StringType.DEFAULT_COLLATION_ID)
```
Comment (stefankandic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ToStringBase.scala:236]: why not use the specified `collationId`?

**Pull request closed without merging**